zipkin/src/main/java/zipkin2/internal/DelayLimiter.java

```diff
@@ -0,0 +1,142 @@
+import java.util.concurrent.Delayed;
+import java.util.concurrent.TimeUnit;
+
+/** Limits invocations of a given context to at most once per period. */
+// this is a dependency-free variant formerly served by an expiring guava cache
+public final class DelayLimiter<C> {
+
+  public static <C> DelayLimiter<C> create() {
+    return new Builder().build();
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+    long ttlNanos = TimeUnit.HOURS.toNanos(1); // legacy default from cassandra
+    int maximumSize = 5 * 1000; // Ex. 5 site tags with cardinality 1000 each
+
+    /**
+     * When {@link #shouldInvoke(Object)} returns true, it will return false until this duration
+     * expires.
+     */
+    public Builder ttl(int ttl) {
+      if (ttl <= 0) throw new IllegalArgumentException("ttl <= 0");
+      this.ttlNanos = TimeUnit.MILLISECONDS.toNanos(ttl);
+      return this;
+    }
+
+    /**
+     * This bounds supressions, useful because contexts can be accidentally unlimited cardinality.
+     */
+    public Builder maxSize(int maximumSize) {
+      if (maximumSize <= 0) throw new IllegalArgumentException("maxSize <= 0");
+      this.maximumSize = maximumSize;
+      return this;
+    }
+
+    public <C> DelayLimiter<C> build() {
+      return new DelayLimiter<>(this);
+    }
+
+    Builder() {
+    }
+  }
+
+  final ConcurrentHashMap<C, Suppression<C>> cache = new ConcurrentHashMap<>();
+  final DelayQueue<Suppression<C>> suppressions = new DelayQueue<>();
+  final long ttlNanos, maximumSize;
+
+  DelayLimiter(Builder builder) {
+    ttlNanos = builder.ttlNanos;
+    maximumSize = builder.maximumSize;
+  }
+
+  /** Returns true if a given context should be invoked. */
+  public boolean shouldInvoke(C context) {
+    cleanupExpiredSuppressions();
+
+    if (cache.containsKey(context)) return false;
+
+    Suppression<C> suppression = new Suppression<>(context, System.nanoTime() + ttlNanos);
+
+    if (cache.putIfAbsent(context, suppression) != null) return false; // lost race
+
+    suppressions.offer(suppression);
+
+    // If we added an entry, it could make us go over the max size. This loops until we are no more
+    // than max size.
+    while (suppressions.size() > maximumSize) {
+      tryClearOneSuppression();
+    }
+
+    return true;
+  }
+
+  public void invalidate(C context) {
+    Suppression<C> suppression = cache.remove(context);
+    if (suppression != null) suppressions.remove(suppression);
+  }
+
+  public void clear() {
+    cache.clear();
+    suppressions.clear();
+  }
+
+  /** This attempts to remove the oldest entry to free up one slot */
+  void tryClearOneSuppression() {
+    // This double-checks that another thread didn't remove the same context we peeked
+    Suppression<C> eldest;
+    if ((eldest = suppressions.peek()) != null && suppressions.remove(eldest)) {
+      // This double-checks that another thread didn't resolve the space problem
```

Comment:
This seems like overkill if the impact is an extra lost suppression or so. It seems like it's safe to remove the while loop above if we loosen the preciseness here which might be safer.

Reply:
I thought so too at first, but then I put in a parallel test and we leaked 12%. The problem this is trying to solve is to make the data structure bounded, which maybe isn't completely clear in the docs..

Reply:
will try a different way before giving up :P

Reply:
yeah I think you are right. the main goal is not to go over size, vs worrying about going under. I will remove the check here and just let it go to 999 in a race condition which is better than the other way!

Reply:
I think I got what you mean now. check me!